finagle-netty4/src/test/scala/com/twitter/finagle/netty4/Netty4SslTest.scala

```diff
@@ -0,0 +1,111 @@
+        pipeline.addLast(new StringEncoder())
+      }, addr, params)
+
+    override protected def newDispatcher(transport: Transport[String, String]): Service[String, String] =
+      new SerialClientDispatcher(transport)
+
+    override protected def copy1(stack: Stack[ServiceFactory[String, String]], params: Params) =
+      copy(clientCert, serverCert, params, stack)
+  }
+}
+
+@RunWith(classOf[JUnitRunner])
+class Netty4SslTest extends FunSuite with Eventually with IntegrationPatience {
+  object StringServerInit extends (ChannelPipeline => Unit) {
+    def apply(pipeline: ChannelPipeline): Unit = {
+      pipeline.addLast("line", new DelimiterBasedFrameDecoder(100, Delimiters.lineDelimiter(): _*))
+      pipeline.addLast("stringDecoder", new StringDecoder(UTF_8))
+      pipeline.addLast("stringEncoder", new StringEncoder(UTF_8))
+    }
+  }
+
+  val serverCert = new SelfSignedCertificate()
+  val clientCert = new SelfSignedCertificate()
+  val allocator = io.netty.buffer.UnpooledByteBufAllocator.DEFAULT
+
+  val server = {
+    val service =
+      new Service[String, String] {
+        override def apply(request: String): Future[String] = {
+          Future.value(
+            Transport.peerCertificate.fold("ERROR")(_ => "OK\n")
+          )
+        }
+      }
+
+    val p = Params.empty +
+      ServerSsl(Some(SslServerConfiguration(clientAuth = ClientAuth.Needed))) +
+      SslServerEngineFactory.Param( new SslServerEngineFactory {
+        override def apply(config: SslServerConfiguration): Engine = {
+          val context = SslContextBuilder.forServer(serverCert.key(), serverCert.cert()).trustManager(clientCert.cert()).build()
+          val engine = context.newEngine(allocator)
+          engine.setNeedClientAuth(true)
+          new Engine(engine)
+        }
+      })
+    Label("test")
+    val listener = Netty4Listener[String, String](StringServerInit, p)
+    val serveTransport = (t: Transport[String, String]) => {
+      if( t.peerCertificate.isEmpty ) throw new IllegalStateException("No peer certificate in transport")
+      new SerialServerDispatcher(t, service)
+    }
+    listener.listen(new InetSocketAddress(InetAddress.getLoopbackAddress, 0))(serveTransport(_))
+  }
+
+
+  test("Peer certificate is available to service") {
+    val addr = server.boundAddress.asInstanceOf[InetSocketAddress]
+
+    val client = new finagle.netty4.Netty4SslTest.Client(clientCert, serverCert)
+      .withTransport.tls(SslClientConfiguration(), new SslClientEngineFactory {
+        override def apply(address: Address, config: SslClientConfiguration): Engine = {
+          val ctx = SslContextBuilder.forClient.keyManager(clientCert.key(), clientCert.cert()).trustManager(serverCert.cert()).build()
+          new Engine(ctx.newEngine(allocator))
+        }
+      })
+      .newService(s"${addr.getHostName}:${addr.getPort}", "client")
+
+    val response = Await.result(client("security is overrated!\n"))
```

Comment:
please add a timeout to the Await, so that if we introduce a regression the test doesn't hang.